File: flink-datastream-1.7.1/src/main/java/source/SmokeLevelSource.java
package source;

import org.apache.flink.streaming.api.functions.source.SourceFunction;
import util.SmokeLevel;

import java.util.Random;

public class SmokeLevelSource implements SourceFunction<SmokeLevel> {
    private boolean running = true;
    @Override
    public void run(SourceContext<SmokeLevel> sourceContext) throws Exception {
        Random rand = new Random();
        while (running) {
            if (rand.nextGaussian() > 0.8) {
                sourceContext.collect(SmokeLevel.HIGH);
            } else {
                sourceContext.collect(SmokeLevel.LOW);
            }
            Thread.sleep(1000);
        }
    }
    @Override
    public void cancel() {
        running = false;
    }
}

File: flink-datastream-1.7.1/src/main/scala/scalalang/source/SmokeLevelSource.scala
package scalalang.source

import org.apache.flink.streaming.api.functions.source.{RichParallelSourceFunction, SourceFunction}
import scalalang.util.SmokeLevel
import scalalang.util.SmokeLevel.SmokeLevel

import scala.util.Random

class SmokeLevelSource extends RichParallelSourceFunction[SmokeLevel] {
  var running = true
  override def run(sourceContext: SourceFunction.SourceContext[SmokeLevel]): Unit = {
    val rand = new Random()
    while (running) {
      if (rand.nextGaussian() > 0.8) {
        sourceContext.collect(SmokeLevel.High)
      } else {
        sourceContext.collect(SmokeLevel.Low)
      }
      Thread.sleep(1000)
    }
  }
  override def cancel(): Unit = {
    running = false
  }
}
